src/GafferSceneUI/ParameterInspector.cpp

```diff
@@ -108,11 +108,22 @@ Gaffer::ValuePlugPtr ParameterInspector::source( const GafferScene::SceneAlgo::H
 
 	if( auto light = runTimeCast<Light>( sceneNode ) )
 	{
-		if( auto optionalPlug = light->parametersPlug()->getChild<OptionalValuePlug>( m_parameter.name ) )
+		if( m_parameter.shader.string().empty() )
 		{
-			return optionalPlug->enabledPlug()->getValue() ? optionalPlug : nullptr;
+			if( auto optionalPlug = light->parametersPlug()->getChild<OptionalValuePlug>( m_parameter.name ) )
+			{
+				return optionalPlug->enabledPlug()->getValue() ? optionalPlug : nullptr;
+			}
+		}
+		/// \todo Remove the need to search for a `ShaderPlug` by adding such a plug to
+		/// `GafferScene::Light` itself.
+		for( const auto &plug : Plug::Range( *light ) )
+		{
+			if( const auto shaderPlug = runTimeCast<ShaderPlug>( plug ) )
+			{
+				return const_cast<ValuePlug *>( shaderPlug->parameterSource( m_parameter ) );
```

Comment:
This is my main concern for the new API - is this `const_cast` a valid solution? The `NetworkBuilder` works with `const` plugs throughout, which does seem right considering what `NetworkBuilder` is meant to do.

Just to see how it would go, I tried making the return plug non-`const`, but the ripple effects of that went all the way up to the `Shader` API. Which doesn't seem like a good thing.

Reply:
I think a `const_cast` is a valid solution, but I don't think this is the place for it. You're right that NetworkBuilder should continue to work with `const` plugs, because it must never modify the graph. And it's also right that `ShaderPlug::parameterSource() const` should return a const result (if the caller only has const access to ShaderPlug, they should only be able to use it to get const access to the rest of the graph). But here we have non-const access to ShaderPlug, so ShaderPlug should also provide us with a non-const overload for `parameterSource()` that returns us a non-const plug.